common/transport/amqp/src/amqp_cbs.ts

```diff
@@ -0,0 +1,323 @@
+    putTokenTimeOutExaminationInterval: number = 10000;
+    timeoutTimer: number;
+}
+
+export class ClaimsBasedSecurityAgent {
+  private static _putTokenSendingEndpoint: string = '$cbs';
+  private static _putTokenReceivingEndpoint: string = '$cbs';
+  private _amqp10Client: amqp10.AmqpClient;
+  private _errorHandler: (err: Error)  => void;
+  private _fsm: machina.Fsm;
+  private _senderLink: SenderLink;
+  private _receiverLink: ReceiverLink;
+  private _putToken: PutTokenStatus = new PutTokenStatus();
+  private _putTokenQueue: {
+    audience: string,
+    token: string,
+    callback: (err?: Error) => void
+  }[];
+
+  constructor(amqp10Client: amqp10.AmqpClient) {
+    this._errorHandler = (err: Error): void => {
+      this._fsm.handle('detach', undefined, err);
+    };
+
+    this._amqp10Client = amqp10Client;
+    /*Codes_SRS_NODE_AMQP_CBS_16_001: [The `constructor` shall instanciate a `SenderLink` object for the `$cbs` endpoint using a custom policy `{encoder: function(body) { return body;}}` which forces the amqp layer to send the token as an amqp value in the body.]*/
+    this._senderLink = new SenderLink(ClaimsBasedSecurityAgent._putTokenSendingEndpoint, { encoder: (body) => body }, this._amqp10Client);
+    this._senderLink.on('error', this._errorHandler);
+
+    /*Codes_SRS_NODE_AMQP_CBS_16_002: [The `constructor` shall instanciate a `ReceiverLink` object for the `$cbs` endpoint.]*/
+    this._receiverLink = new ReceiverLink(ClaimsBasedSecurityAgent._putTokenReceivingEndpoint, null, this._amqp10Client);
+    this._receiverLink.on('error', this._errorHandler);
+
+    this._putTokenQueue = [];
+
+    this._fsm = new machina.Fsm({
+      initialState: 'detached',
+      states: {
+        detached: {
+          _onEnter: (callback, err) => {
+            let tokenOperation = this._putTokenQueue.shift();
+            while (tokenOperation) {
+              tokenOperation.callback(err);
+              tokenOperation = this._putTokenQueue.shift();
+            }
+
+            /*Codes_SRS_NODE_AMQP_CBS_16_006: [If given as an argument, the `attach` method shall call `callback` with a standard `Error` object if any link fails to attach.]*/
+            if (callback) {
+              callback(err);
+            }
+          },
+          attach: (callback) => {
+            this._fsm.transition('attaching', callback);
+          },
+          detach: (callback) => { if (callback) callback(); },
+          forceDetach: () => { return; },
+          putToken: (audience, token, callback) => {
+            this._putTokenQueue.push({
+              audience: audience,
+              token: token,
+              callback: callback
+            });
+            this._fsm.transition('attaching');
+          }
+         },
+        attaching: {
+          _onEnter: (callback) => {
+            this._senderLink.attach((err) => {
+              if (err) {
+                this._fsm.transition('detached', callback, err);
```

Comment:
>this._fsm.transition('detached', callback, err); [](start = 16, length = 48)

I can see why you're going straight into detached here, but it might be worth going through the detaching state just to future proof your code.  (quick eyes might not notice that a failure on _senderLink goes into a different state than the failure on _receiverLink)

Reply:
yeah that's not a bad idea..

---
In reply to: [139538455](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139538455) [](ancestors = 139538455)